core/chains/evm/logpoller/log_poller.go

```diff
@@ -0,0 +1,380 @@
+	finalityDepth     int64         // finality depth is taken to mean that block (head - finality) is finalized
+	backfillBatchSize int64         // batch size to use when backfilling finalized logs
+
+	filterMu  sync.Mutex
+	addresses map[common.Address]struct{}
+	topics    map[int]map[common.Hash]struct{}
+
+	replay chan int64
+	ctx    context.Context
+	cancel context.CancelFunc
+	done   chan struct{}
+}
+
+func NewLogPoller(orm *ORM, ec client.Client, lggr logger.Logger, pollPeriod time.Duration, finalityDepth, backfillBatchSize int64) *LogPoller {
+	return &LogPoller{
+		ec:                ec,
+		orm:               orm,
+		lggr:              lggr,
+		replay:            make(chan int64),
+		done:              make(chan struct{}),
+		pollPeriod:        pollPeriod,
+		finalityDepth:     finalityDepth,
+		backfillBatchSize: backfillBatchSize,
+		addresses:         make(map[common.Address]struct{}),
+		topics:            make(map[int]map[common.Hash]struct{}),
+	}
+}
+
+// MergeFilter will update the filter with the new topics and addresses.
+// Clients may chose to MergeFilter and then replay in order to ensure desired logs are present.
+func (lp *LogPoller) MergeFilter(topics []common.Hash, address common.Address) {
+	lp.filterMu.Lock()
+	defer lp.filterMu.Unlock()
+	lp.addresses[address] = struct{}{}
+	// [[A, B], [C]] + [[D], [], [E]] = [[A, B, D], [C], [E]]
+	for i := 0; i < len(topics); i++ {
+		if lp.topics[i] == nil {
+			lp.topics[i] = make(map[common.Hash]struct{})
+		}
+		lp.topics[i][topics[i]] = struct{}{}
+	}
+}
+
+func (lp *LogPoller) filterAddresses() []common.Address {
+	lp.filterMu.Lock()
+	defer lp.filterMu.Unlock()
+	var addresses []common.Address
+	for addr := range lp.addresses {
+		addresses = append(addresses, addr)
+	}
+	sort.Slice(addresses, func(i, j int) bool {
+		return bytes.Compare(addresses[i][:], addresses[j][:]) < 0
+	})
+	return addresses
+}
+
+func (lp *LogPoller) filterTopics() [][]common.Hash {
+	lp.filterMu.Lock()
+	defer lp.filterMu.Unlock()
+	var topics [][]common.Hash
+	for idx := 0; idx < len(lp.topics); idx++ {
+		var topicPosition []common.Hash
+		for topic := range lp.topics[idx] {
+			topicPosition = append(topicPosition, topic)
+		}
+		sort.Slice(topicPosition, func(i, j int) bool {
+			return bytes.Compare(topicPosition[i][:], topicPosition[j][:]) < 0
+		})
+		topics = append(topics, topicPosition)
+	}
+	return topics
+}
+
+// Replay signals that the poller should resume from a new block.
+func (lp *LogPoller) Replay(fromBlock int64) {
+	lp.replay <- fromBlock
+}
+
+func (lp *LogPoller) Start(parentCtx context.Context) error {
+	return lp.StartOnce("LogPoller", func() error {
+		ctx, cancel := context.WithCancel(parentCtx)
+		lp.ctx = ctx
+		lp.cancel = cancel
+		go lp.run()
+		return nil
+	})
+}
+
+func (lp *LogPoller) Close() error {
+	return lp.StopOnce("LogPoller", func() error {
+		lp.cancel()
+		<-lp.done
+		return nil
+	})
+}
+
+func (lp *LogPoller) Ready() error {
+	return nil
```

Comment:
Is it a good idea to return "Not ready", till we are able to atleast get a correct start value in the poller?
With that, other dependencies can wait processing their stuff till they know this poller is actually live.

Reply:
I think for log based jobs, all they really care about is the latest block the poller is on, so we should expose that so they know if its making progress. Ready/healthy are orthogonal, as they impact only out of process traffic/node restarts